SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityOfferHolder.java

```diff
@@ -124,7 +124,20 @@ public void launchTasks(SchedulerDriver driver) {
       LOG.trace("Launching {} mesos task: {}", task.getTaskId(), MesosUtils.formatForLogging(task.getMesosTask()));
     }
 
-    Status initialStatus = driver.launchTasks(offers.stream().map(Protos.Offer::getId).collect(Collectors.toList()), toLaunch);
+    // At this point, `currentResources` contains a list of unused resources, because we subtracted out the required resources of every task we accepted.
+    // Let's try and reclaim offers by trying to pull each offer's list of resources out of the combined pool of leftover resources.
+    List<Offer> neededOffers = offers.stream().filter(o -> {
+      List<Resource> remainingAfterSavingOffer = MesosUtils.subtractResources(currentResources, o.getResourcesList());
+      if (MesosUtils.allResourceCountsNonNegative(remainingAfterSavingOffer)) {
+        cache.cacheOffer(driver, System.currentTimeMillis(), o); // TODO: do we need this timestamp to be something specific, e.g. the time at which we got the offer originally?
```

Comment:
keep in mind here, the offer may already be from the cache to start with, we might need to be returning the offer rather than caching it here